src/Npgsql/PoolManager.cs

```diff
@@ -13,20 +15,49 @@ namespace Npgsql;
 /// </remarks>
 static class PoolManager
 {
-    internal static ConcurrentDictionary<string, NpgsqlDataSource> Pools { get; } = new();
+    static ConcurrentDictionary<string, NpgsqlDataSource> _pools = new();
+
+    // Keep the IEnumerable API for testing purposes but remove the rest to avoid careless (unlocked) access to the
+    // private dictionary.
+    internal static IEnumerable<KeyValuePair<string, NpgsqlDataSource>> Pools => _pools.ToImmutableDictionary();
 
     internal static void Clear(string connString)
     {
-        // TODO: Actually remove the pools from here, #3387 (but be careful of concurrency)
-        if (Pools.TryGetValue(connString, out var pool))
-            pool.Clear();
+        // ReSharper disable once InconsistentlySynchronizedField
+        var pools = _pools;
```

Comment:
Is there any reason not to put this entire method in the lock? I think it's simpler/safer to always take this lock when mutating the pools dictionary (which is a rare event).

Reply:
It isn't necessary since we're only mutating the content of the concurrent dictionary (which is thread safe) but not the reference to it.

The only reason (that I can see) to completely put both, the entire `Clear()` and the entire `ClearAll()` method in a lock would be to avoid accidentally running out of backend connections because a newly added pool could already be filling up while the old one is still cooling down (in `DataSource.Dispose()`).

I agree that `Clear()`/`ClearAll()` should rarely happen, so perhaps let's just do this.

Reply:
> I agree that Clear()/ClearAll() should rarely happen, so perhaps let's just do this.

On the other hand, disposing a pool can really take quite a while (waiting until all the connections are either back to idle or broken) and not allowing to add a new pool with the same connection string back to the pool while this happens might really hurt.

Reply:
> On the other hand, disposing a pool can really take quite a while

Unless I'm mistaken, disposing a pool doesn't wait until all connections are idle or broken - it just closes all current idle connections (which should generally be instantaneous), disposes the password provider timer, and sets state to make sure subsequent Open attempts throw.

I agree it isn't strictly necessary to put a lock around everything - but I still think it's a good idea. There's no perf aspect here (again, since Clear/ClearAll is rare and isn't perf-sensitive), so keeping things simple and very clear should be the priority IMHO; and just locking all all dictionary modifications makes it easier to reason about everything (i.e. less lock-free is always better).

But if you have strong feelings here I'm OK with not doing it.

Reply:
> Unless I'm mistaken, disposing a pool doesn't wait until all connections are idle or broken - it just closes all current idle connections (which should generally be instantaneous), disposes the password provider timer, and sets state to make sure subsequent Open attempts throw.

You're right. I missed that it's iterating while **`_idleCount`** is greater than 0.
Yeah, then let's do it the simple way.